### PR TITLE
2 e2e test with deposit checking and MGN checking

### PR DESCRIPTION
**Pull request by josojo**
the e2e test revealed two errors:

- in function triggerMGNunlockAndClaimTokens()
we first want to claim token and then do the unlocking, as the claiming will add some more MGN tokens

- totaldeposit was not updated during withdraw

### CODE REVIEW

File: test/utilities.js [hunk 1]
```diff
@@ -1,6 +1,8 @@
-const waitForNBlocks = async function(numBlocks, authority) {
+
```
Comment (fleupold) [inline test/utilities.js]: Can we clean this up? `timeTravel` and `increaseTimeBy` should do the same thing (same for `timestamp` and `getTime`). If they are not the same thing their difference should be made clear.

Generally it would be nice if this kind of utility could be provided by a library. It feels like we create this type of helper in all our repos.

File: test/dx_e2e.js [hunk 2]
```diff
@@ -1,56 +1,115 @@
...
     await depositToken.approve(dx.address, oneEth)
 
     await secondaryTokenMock.givenAnyReturnBool(true)
 
-    await instance.deposit(10)
-    assert.equal(await instance.numberOfParticipations.call(accounts[0]), 1)
+    await instance1.deposit(10)
+    assert.equal(await instance1.numberOfParticipations.call(accounts[0]), 1)
 
     //frist tokenPair needs to be funded first...
```
Comment (fleupold) [inline test/dx_e2e.js]:
```suggestion
    //first tokenPair needs to be funded first...
```

File: test/dx_e2e.js [hunk 3]
```diff
@@ -1,56 +1,115 @@
...
-contract("Trading", (accounts) => {
+contract("e2e - tests", (accounts) => {
   it("adds a particpation and deposits them into first auction", async () => {
     const secondaryTokenMock = await MockContract.new()
     const mgnTokenMock = await MockContract.new()
     const dxProxy = await DXProxy.deployed()
     const dx = await DX.at(dxProxy.address)
-    const erc20 = await ERC20.new()
+    const mgnToken = await TokenFRT.at(await dx.frtToken.call())
     const depositToken = await EtherToken.at(await dx.ethToken.call())
```
Comment (fleupold) [inline test/dx_e2e.js]: In this context probably also token1/token2

File: test/dx_e2e.js [hunk 4]
```diff
@@ -1,56 +1,115 @@
...
 
 const truffleAssert = require("truffle-assertions")
-const { waitForNBlocks, timeTravel, timestamp } = require("./utilities")
+const { waitForNBlocks, timeTravel, timestamp, increaseTimeBy } = require("./utilities")
 const BN = web3.utils.BN
 
 
-contract("Trading", (accounts) => {
+contract("e2e - tests", (accounts) => {
   it("adds a particpation and deposits them into first auction", async () => {
```
Comment (fleupold) [inline test/dx_e2e.js]: this description seems outdated

File: test/dx_e2e.js [hunk 5]
```diff
@@ -1,56 +1,115 @@
...
-    //  await instance.participateInAuction()
+    await instance1.triggerMGNunlockAndClaimTokens()
+    
+    console.log("could withdraw balance")
+    const balBefore = await depositToken.balanceOf.call(accounts[0]);
+    await instance1.withdrawDeposit()
+    const balAfter = await depositToken.balanceOf.call(accounts[0]);
 
-  	  // assert.equal(await dx.sellerBalance.call(DxMgnPool.address, 0), 10)*/
+    assert.equal(balAfter.sub(balBefore).toString(), 9);
```
Comment (fleupold) [inline test/dx_e2e.js]: Let's assert on the MGN as well?

Comment (fleupold) [inline test/dx_e2e.js]: Why exactly 9 difference?

File: test/dx_e2e.js [hunk 6]
```diff
@@ -1,56 +1,115 @@
...
+    await dx.postBuyOrder(secondaryTokenMock.address, depositToken.address, 1, oneGwei.toString())
+
+    assert.equal(await dx.getAuctionIndex(depositToken.address, secondaryTokenMock.address),2)
+
+    //claim funds  and post into next auction auction with more
+    await dx.claimSellerFunds(depositToken.address, secondaryTokenMock.address, accounts[0], 1)
+    await dx.claimSellerFunds(secondaryTokenMock.address, depositToken.address, accounts[0], 1)
+    await dx.claimBuyerFunds(depositToken.address, secondaryTokenMock.address, accounts[0], 1)
+    await dx.claimBuyerFunds(secondaryTokenMock.address, depositToken.address, accounts[0], 1)
+    await dx.postSellOrder(depositToken.address, secondaryTokenMock.address, 0, oneGwei.toString())
```
Comment (fleupold) [inline test/dx_e2e.js]: Is this required? Who will do this in our model?

File: test/dx_e2e.js [hunk 7]
```diff
@@ -1,56 +1,115 @@
...
+    await dx.postSellOrder(secondaryTokenMock.address, depositToken.address, 0, oneGwei.toString())
+
+    //close the auction and pool paritcpation in next one
+    await waitUntilPriceIsXPercentOfPreviousPrice(depositToken, secondaryTokenMock, 1);
+    await dx.postBuyOrder(depositToken.address, secondaryTokenMock.address, 2, oneGwei.toString())
```
Comment (fleupold) [inline test/dx_e2e.js]: I don't understand why there seems to be an extra auction, in which we don't participate. Is this just so we test that this works or is there more to it? Could you document it more? I'm confused, because it seems to happen below as well.

File: test/utilities.js (hunk 1 above)
Comment (josojo) [reply test/utilities.js]: Cleaned it up. Yeah, we also discussed this library approach in one of your weekly smart contracts.

Unfortunately, it never got any traction...

File: test/dx_e2e.js (hunk 5 above)
Comment (josojo) [reply test/dx_e2e.js]: Yeah, I will add this next...

Comment (josojo) [reply test/dx_e2e.js]: We should get 10 out of it, but the second auction the price is not 1, it's 0.999999 and then it rounds down from 10 to 9.

File: test/dx_e2e.js (hunk 6 above)
Comment (josojo) [reply test/dx_e2e.js]: Other bots have to post sell orders and make sure that prices do not fall too much. These sell orders clear the auction

And I am just doing the claiming, as I want the tokens back for the next auction

File: test/dx_e2e.js (hunk 7 above)
Comment (josojo) [reply test/dx_e2e.js]: I all the time have to run the auction WETH-GNO and GNO-WETH. I just can't have a single one of them.

File: contracts/DxMgnPool.sol [hunk 8]
```diff
@@ -129,12 +130,12 @@ contract DxMgnPool is Ownable {
...
         );
         
-        mgnToken.unlockTokens();
-        
         (address sellToken, address buyToken) = buyAndSellToken();
         dx.claimSellerFunds(buyToken, sellToken, address(this), lastParticipatedAuctionIndex);
-        uint amount = dx.balances(address(depositToken), address(this));
-        dx.withdraw(address(depositToken), amount);
+
+        mgnToken.unlockTokens();
```
Comment (fleupold) [inline contracts/DxMgnPool.sol]: Can you please make this its own PR? It makes it easy to track changes and if necessary revert them individually.

File: contracts/DxMgnPool.sol [hunk 9]
```diff
@@ -197,4 +198,6 @@ contract DxMgnPool is Ownable {
             return (address(secondaryToken), address(depositToken)); 
         }
     }
+
+    event NewDeposit(uint amount, address sender);
```
Comment (fleupold) [inline contracts/DxMgnPool.sol]: Do we really want this? Should also be separate PR then, please

File: test/dx_e2e.js [hunk 10]
```diff
@@ -1,56 +1,145 @@
...
 const BN = web3.utils.BN
 
 
-contract("Trading", (accounts) => {
-  it("adds a particpation and deposits them into first auction", async () => {
-    const secondaryTokenMock = await MockContract.new()
-    const mgnTokenMock = await MockContract.new()
+contract("e2e - tests", (accounts) => {
+  it("e2e tests for deposits: 1 deposit - 2x trading - withdraw", async () => {
+    const  token_2 = await TokenGNO.new("1111111111111111111111")
```
Comment (fleupold) [inline test/dx_e2e.js]: what is this magic number?

File: test/utilities.js [hunk 11]
```diff
@@ -1,25 +1,59 @@
...
+      if (!err) return accept(resp)
+
+      return rej(err)
+    })
+  })
+}
 
 
 module.exports = {
+  mineCurrentBlock,
```
Comment (fleupold) [inline test/utilities.js]: `mineCurrentBlock` doesn't need to be exported.

Comment (fleupold) [reply test/utilities.js]: still there

**Comment by fleupold** [comment]
Otherwise good to merge once checks pass
